### PR TITLE
- Add async interface to put

### PR DESCRIPTION
**Pull request by baldersheim**
- Use MessageTracker for keeping context.
- implement putAsync, but still use it synchronously.

@vekterli PR

### CODE REVIEW

File: searchcore/src/vespa/searchcore/proton/common/feedtoken.h [hunk 1]
```diff
@@ -15,36 +15,53 @@ typedef std::unique_ptr<storage::spi::Result> ResultUP;
...
+    const storage::spi::Result &getResult() { return *_result; }
+protected:
+    void ack();
+private:
+    ITransport           &_transport;
+    ResultUP              _result;
+    bool                  _documentWasFound;
+    std::atomic<bool>     _alreadySent;
+};
+class OwningState : public State {
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/common/feedtoken.h:45]: Consider adding class description.

File: searchcore/src/vespa/searchcore/proton/persistenceengine/transport_latch.cpp [hunk 2]
```diff
@@ -8,10 +8,50 @@ using storage::spi::Result;
...
+    if (_lock) {
+        std::lock_guard<std::mutex> guard(*_lock);
+        mergeWithLock(std::move(result), documentWasFound);
+    } else {
+        mergeWithLock(std::move(result), documentWasFound);
+    }
+}
+
+void
+TransportMerger::mergeWithLock(ResultUP result, bool documentWasFound) {
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/persistenceengine/transport_latch.cpp:34]: Consider adding a reference to the lock to ensure that we don't call this when the lock is not taken.

File: searchcore/src/vespa/searchcore/proton/persistenceengine/transport_latch.h [hunk 3]
```diff
@@ -13,18 +13,33 @@ namespace proton {
  * Implementation of FeedToken::ITransport for handling the async reply for an operation.
  * Uses an internal count down latch to keep track the number of outstanding replies.
  */
-class TransportLatch : public feedtoken::ITransport {
-private:
+
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/persistenceengine/transport_latch.h:16]: Is class comment still correct?

File: searchcore/src/vespa/searchcore/proton/persistenceengine/transport_latch.h [hunk 4]
```diff
@@ -13,18 +13,33 @@ namespace proton {
...
+    ~TransportMerger() override;
+    void mergeResult(ResultUP result, bool documentWasFound);
+    virtual void completeIfDone() { } // Called with lock held if necessary on every merge
+    ResultUP  _result;
+
+private:
+    void mergeWithLock(ResultUP result, bool documentWasFound);
+    std::unique_ptr<std::mutex> _lock;
+};
+class TransportLatch : public TransportMerger {
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/persistenceengine/transport_latch.h:32]: Please add class comment.

**Review by geirst — COMMENTED** [review]
Some comments to consider (I have only looked at searchcore changes)

**Comment by geirst** [comment]
Also consider if testing of TransportLatch and AsyncTranportContext is good enough. They are implicitly tested in persistenceengine_test.cpp.

File: storage/src/vespa/storage/persistence/persistenceutil.cpp [hunk 5]
```diff
@@ -14,22 +14,69 @@ namespace {
...
+MessageTracker::setMetric(FileStorThreadMetrics::Op& metric) {
+    metric.count.inc();
+    _metric = &metric;
 }
 
 MessageTracker::~MessageTracker()
 {
     if (_reply.get() && _reply->getResult().success()) {
-        _metric.latency.addValue(_timer.getElapsedTimeAsDouble());
+        _metric->latency.addValue(_timer.getElapsedTimeAsDouble());
```
Comment (vekterli) [inline storage/src/vespa/storage/persistence/persistenceutil.cpp:58]: Is it valid behavior with the current code that a tracker still has a reply upon destruction? If not, when is the latency metric updated? Should it be moved to `sendReply()`?

File: storage/src/vespa/storage/persistence/persistenceutil.cpp [hunk 6]
```diff
@@ -14,22 +14,69 @@ namespace {
...
+            if (getReply().getResult().success()) {
+                _env.setBucketInfo(*this, _bucketLock->getBucket());
+            }
+        }
+        LOG(spam, "Sending reply up: %s %" PRIu64,
+            getReply().toString().c_str(), getReply().getMsgId());
+        _env._fileStorHandler.sendReply(std::move(_reply));
+    } else {
+        if ( ! _context.getTrace().getRoot().isEmpty()) {
+            _msg->getTrace().getRoot().addChild(_context.getTrace().getRoot());
```
Comment (vekterli) [inline storage/src/vespa/storage/persistence/persistenceutil.cpp:78]: Does it make sense to append a trace to a message that has no reply associated with it?

**Review by vekterli — APPROVED** [review]
:+1: some questions and comments for consideration

File: searchcore/src/vespa/searchcore/proton/common/feedtoken.h (hunk 1 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/common/feedtoken.h:45]: Fixed

File: searchcore/src/vespa/searchcore/proton/persistenceengine/transport_latch.h (hunk 4 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/persistenceengine/transport_latch.h:32]: Fixed

File: searchcore/src/vespa/searchcore/proton/persistenceengine/transport_latch.h (hunk 3 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/persistenceengine/transport_latch.h:16]: Now it is.

File: searchcore/src/vespa/searchcore/proton/persistenceengine/transport_latch.cpp (hunk 2 above)
Comment (baldersheim) [reply searchcore/src/vespa/searchcore/proton/persistenceengine/transport_latch.cpp:34]: I did not do that sine the lock is optional.

File: persistence/src/vespa/persistence/spi/operationcomplete.h [hunk 7]
```diff
@@ -0,0 +1,22 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply persistence/src/vespa/persistence/spi/operationcomplete.h:1]: Fixed

File: persistence/src/vespa/persistence/spi/persistenceprovider.cpp [hunk 8]
```diff
@@ -1,9 +1,35 @@
...
 #include "persistenceprovider.h"
+#include <future>
 
 namespace storage::spi {
 
 PersistenceProvider::~PersistenceProvider() = default;
 
+class CatchResult : public OperationComplete {
+public:
+    std::future<Result::UP> waitResult() {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply persistence/src/vespa/persistence/spi/persistenceprovider.cpp:12]: Fixed

File: storage/src/vespa/storage/persistence/persistenceutil.cpp (hunk 6 above)
Comment (baldersheim) [reply storage/src/vespa/storage/persistence/persistenceutil.cpp:78]: I do not know, I just tried to preserve current behavior.

**Pull request merged**